File: XWalkView/XWalkView/XWalkExtension.m
// Copyright (c) 2014 Intel Corporation. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "XWalkExtension.h"

#import "Invocation.h"
#import "XWalkView/XWalkView-Swift.h"

@interface XWalkExtension()

@property(nonatomic, weak) XWalkChannel* channel;
@property(nonatomic, assign) NSInteger instance;

- (id)objectForKeyedSubscript:(NSString *)key;
- (void)setObject:(id)obj forKeyedSubscript:(id <NSCopying>)key;

@end

@implementation XWalkExtension {
    BOOL _sync;
}

- (NSString*)namespace {
    return self.channel ? self.channel.namespace : nil;
}

- (void)invokeNativeMethod:(NSString *)name arguments:(NSArray *)args {
    SEL selector = [self.channel.mirror getMethod:name];
    if (!selector) {
        NSLog(@"ERROR: Method '%@' is undefined in class '%@'.", name, NSStringFromClass(self.class));
        return;
    }

    ReturnValue *result = [Invocation call:self selector:selector arguments:args];
    if (result.isBool && result.boolValue == YES)
        [self releaseArguments:((NSNumber *)args[0]).unsignedIntValue];
}

- (void)setNativeProperty:(NSString *)name value:(id)value {
    SEL selector = [self.channel.mirror getSetter:name];
    if (!selector) {
        if ([self.channel.mirror hasProperty:name])
            NSLog(@"ERROR: Property '%@' is readonly.", name);
        else
            NSLog(@"ERROR: Property '%@' is undefined in class '%@'.", name, NSStringFromClass(self.class));
        return;
    }

    _sync = NO;
    [Invocation call:self selector:selector arguments:@[value]];
    _sync = YES;
}

- (NSString*)didGenerateStub:(NSString*)stub {
    NSBundle* bundle = [NSBundle bundleForClass:self.class];
    NSString* name = NSStringFromClass(self.class);
    if (name.pathExtension.length) {
        name = name.pathExtension;
    }
    NSString* path = [bundle pathForResource:name ofType:@"js"];
    if (path) {
        NSString* content = [NSString stringWithContentsOfFile:path encoding:NSUTF8StringEncoding error:nil];
        if (content)
            return [stub stringByAppendingString:content];
    }
    return stub;
}

- (void)didBindExtension:(XWalkChannel*)channel instance:(NSInteger)instance {
    self.channel = channel;
    self.instance = instance;

    NSEnumerator* enumerator = [self.channel.mirror.allProperties objectEnumerator];
    NSString *name;
    while (name = [enumerator nextObject]) {
        NSString *key = NSStringFromSelector([self.channel.mirror getGetter:name]);
        [self addObserver:self forKeyPath:key options:NSKeyValueObservingOptionNew context:(__bridge void *)self];
        if (instance)
            [self setJavaScriptProperty:name value:self[name]];
    }
    _sync = YES;
}

- (void)didUnbindExtension {
    NSEnumerator* enumerator = [self.channel.mirror.allProperties objectEnumerator];
    id name;
    while (name = [enumerator nextObject]) {
        NSString *key = NSStringFromSelector([self.channel.mirror getGetter:name]);
        [self removeObserver:self forKeyPath:key context:(__bridge void *)self];
    }

    self.channel = nil;
    self.instance = 0;
}

- (void)setJavaScriptProperty:(NSString*)name value:(id)value {
    NSString* json = nil;
    if (value == nil || value == NSNull.null) {
        json = @"null";
    } else if ([value isKindOfClass:NSString.class]) {
        json = [NSString stringWithFormat:@"'%@'", (NSString*)value];
    } else if ([value isKindOfClass:NSNumber.class]) {
        json = [NSString stringWithFormat:@"%@", value];
    } else {
        NSError* error = nil;
        NSData* data = [NSJSONSerialization dataWithJSONObject:value options:NSJSONWritingPrettyPrinted error:&error];
        if (error) {
            NSLog(@"ERROR: Failed to generate json string from value object.");
            return;
        }
        json = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    }
    NSString *script = [NSString stringWithFormat:@"%@.properties['%@'] = %@;",
            self.instance ? [self.namespace stringByAppendingFormat:@"[%zd]", self.instance] : self.namespace, name, json];
    [self evaluateJavaScript:script];
}

- (void)invokeCallback:(UInt32)callbackId key:(NSString*)key, ... {
    NSMutableArray *args = [NSMutableArray new];
    va_list ap;
    id arg;
    va_start(ap, key);
    while ((arg = va_arg(ap, id)) != nil) {
        [args addObject:arg];
    }
    va_end(ap);
    [self invokeJavaScript:@".invokeCallback", [NSNumber numberWithInteger:callbackId], key ?: NSNull.null, args, nil];
}

- (void)invokeCallback:(UInt32)callbackId key:(NSString*)key arguments:(NSArray*)arguments {
    [self invokeJavaScript:@".invokeCallback", [NSNumber numberWithInteger:callbackId], key ?: NSNull.null, arguments, nil];
}

- (void)releaseArguments:(UInt32)callId {
    [self invokeJavaScript:@".releaseArguments", [NSNumber numberWithUnsignedInteger:callId], nil];
}

- (void)invokeJavaScript:(NSString*)function, ... {
    NSMutableArray *args = [NSMutableArray new];
    va_list ap;
    id arg;
    va_start(ap, function);
    while ((arg = va_arg(ap, id)) != nil) {
        [args addObject:arg];
    }
    [self invokeJavaScript:function arguments:args];
    va_end(ap);
}

- (void)invokeJavaScript:(NSString*)function arguments:(NSArray*)arguments {
    NSMutableString* script = [NSMutableString stringWithString:function];
    NSMutableString* this = [NSMutableString stringWithString:@"null"];
    if ([script characterAtIndex:0] == '.') {
        // Invoke a method of this object
        [this setString:self.namespace];
        if (self.instance)
            [this appendFormat:@"[%zd]", self.instance];
        [script insertString:this atIndex:0];
    }

    NSString* json = @"[]";
    if (arguments != nil) {
        NSError* error;
        NSData* data = [NSJSONSerialization dataWithJSONObject:arguments options:NSJSONWritingPrettyPrinted error:&error];
        if (error) {
            NSLog(@"ERROR: Failed to generate json string from arguments object.");
            return;
        }
        json = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    }
    [script appendString:[NSString stringWithFormat:@".apply(%@, %@);", this, json]];
    [self evaluateJavaScript:script];
}

- (void)evaluateJavaScript:(NSString*)string {
    [self.channel evaluateJavaScript:string completionHandler:^void(id obj, NSError* err) {
        if (err) {
            NSLog(@"ERROR: Failed to execute script, %@\n------------\n%@\n------------", err, string);
        }
    }];
}

- (void)evaluateJavaScript:(NSString*)string onSuccess:(void(^)(id))onSuccess onError:(void(^)(NSError*))onError {
    [self.channel evaluateJavaScript:string completionHandler:^void(id obj, NSError*err) {
        err ? onSuccess(obj) : onError(err);
    }];
}

- (id)objectForKeyedSubscript:(NSString *)key {
    SEL selector = [self.channel.mirror getGetter:key];
    if (selector) {
        ReturnValue* result = [Invocation call:self selector:selector arguments:nil];
        if (result.isObject || result.isNumber)
            return result.object ?: result.number;
        else
            [NSException raise:@"PropertyError" format:@"Type of property '%@' is unknown.", key];
    } else {
        [NSException raise:@"PropertyError" format:@"Property '%@' is undefined.", key];
    }
    return nil;
}

- (void)setObject:(id)obj forKeyedSubscript:(id <NSCopying>)key {
    NSString* name = (NSString*)key;
    SEL selector = [self.channel.mirror getSetter:name];
    if (selector) {
        if (!_sync)
            [self setJavaScriptProperty:name value:obj];
        [Invocation call:self selector:selector arguments:obj ?: NSNull.null];
    } else if ([self.channel.mirror hasProperty:name]) {
        [NSException raise:@"PropertyError" format:@"Property '%@' is readonly.", name];
    } else {
        [NSException raise:@"PropertyError" format:@"Property '%@' is undefined.", name];
    }
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
    // FIXME: This class should not know the mapping between selector and property name.
    if (context == (__bridge void *)(self) && _sync)
        [self setJavaScriptProperty:[keyPath substringFromIndex:7] value:change[NSKeyValueChangeNewKey]];
}

@end

File: XWalkView/XWalkView/XWalkExtension.swift
// Copyright (c) 2014 Intel Corporation. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Foundation

public class XWalkExtension : NSObject, XWalkDelegate {
    private weak var _channel: XWalkChannel!
    private var _instance: Int = 0

    public final weak var channel: XWalkChannel! { return _channel }
    public final var instance: Int { return _instance }
    public final var namespace: String { return _channel.namespace }

    public func didGenerateStub(stub: String) -> String {
        let bundle : NSBundle = NSBundle(forClass: self.dynamicType)
        var name = self.dynamicType.description()
        name = name.pathExtension.isEmpty ? name : name.pathExtension
        if let path = bundle.pathForResource(name, ofType: "js") {
            var error: NSError?
            if let content = NSString(contentsOfFile: path, encoding: UInt(NSUTF8StringEncoding), error: &error) {
                return "\(stub)\n\(content)"
            } else {
                NSException.raise("EncodingError", format: "'%@.js' should be UTF-8 encoding.", arguments: getVaList([name]))
            }
        }
        return stub
    }
    public func didBindExtension(channel: XWalkChannel, instance: Int) {
        _channel = channel
        _instance = instance

        if instance != 0 {
            for name in _channel.mirror.allMembers {
                if _channel.mirror.hasProperty(name) {
                    setProperty(name, value: self[name])
                }
            }
        }
    }

    internal func setProperty(name: String, value: AnyObject?) {
        // TODO: check type
        var val: AnyObject = value ?? NSNull()
        if val.isKindOfClass(NSString.classForCoder()) {
            val = NSString(format: "'\(val as String)'")
        }
        let json = JSON(val).toString()
        let script = "\(_channel.namespace)" + (_instance != 0 ? "[\(instance)]" : "") + ".properties['\(name)'] = \(json);"
        evaluateJavaScript(script)
    }
    public func invokeCallback(id: UInt32, key: String? = nil, arguments: [AnyObject] = []) {
        let args = [NSNumber(unsignedInt: id), key ?? NSNull(), arguments]
        invokeJavaScript(".invokeCallback", arguments: args)
    }
    public func invokeCallback(id: UInt32, index: UInt32, arguments: [AnyObject] = []) {
        let args = [NSNumber(unsignedInt: id), NSNumber(unsignedInt: index), arguments]
        invokeJavaScript(".invokeCallback", arguments: args)
    }
    public func releaseArguments(callid: UInt32) {
        invokeJavaScript(".releaseArguments", arguments: [NSNumber(unsignedInt: callid)])
    }
    public func invokeJavaScript(function: String, arguments: [AnyObject] = []) {
        var script = function
        var this = "null"
        if script[script.startIndex] == "." {
            // Invoke a method of this object
            this = _channel.namespace + (_instance != 0 ? "[\(instance)]" : "")
            script = this + function
        }
        script += ".apply(\(this), \(JSON(arguments).toString()));"
        evaluateJavaScript(script)
    }
    public func evaluateJavaScript(string: String) {
        _channel.evaluateJavaScript(string, completionHandler: { (obj, err)->Void in
            if err != nil {
                println("ERROR: Failed to execute script, \(err)\n------------\n\(string)\n------------")
            }
        })
    }
    public func evaluateJavaScript(string: String, onSuccess: ((AnyObject!)->Void)?, onError: ((NSError!)->Void)?) {
        _channel.evaluateJavaScript(string, completionHandler: { (obj, err)->Void in
            err == nil ? onSuccess?(obj) : onError?(err)
            return    // To make compiler happy
        })
    }

    public subscript(name: String) -> AnyObject? {
        get {
            let selector = _channel.mirror.getGetter(name)
            if selector != nil {
                let result = Invocation.call(self, selector: selector, arguments: nil)
                if let obj: AnyObject = result.object ?? result.number {
                    return obj
                } else if !(result.object is NSNull) {
                    NSException.raise("PropertyError", format: "Type of property '%@' is unknown.", arguments: getVaList([name]))
                }
            } else {
                NSException.raise("PropertyError", format: "Property '%@' is not defined.", arguments: getVaList([name]))
            }
            return nil
        }
        set(value) {
            let selector = _channel.mirror.getSetter(name)
            if selector != nil {
                if channel.mirror.getOriginalSetter(name) == nil {
                    setProperty(name, value: value)
                }
                Invocation.call(self, selector: selector, arguments: [value ?? NSNull()])
            } else if _channel.mirror.hasProperty(name) {
                NSException.raise("PropertyError", format: "Property '%@' is readonly.", arguments: getVaList([name]))
            } else {
                NSException.raise("PropertyError", format: "Property '%@' is not defined.", arguments: getVaList([name]))
            }
        }
    }
}
